internal/util/aws_metadata.go

```diff
@@ -0,0 +1,206 @@
+	err := setConfigString(&kms.namespace, args.Config, args.Namespace)
+	if err != nil {
+		return nil, err
+	}
+	err = setConfigString(&kms.secretName, args.Config, awsSecretNameKey)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	} else if errors.Is(err, errConfigOptionMissing) {
+		kms.secretName = awsMetadataDefaultSecretsName
+	}
+	err = setConfigString(&kms.region, args.Config, awsRegionKey)
+	if err != nil {
+		return nil, err
+	}
+
+	// read the Kubenetes Secret with credentials
+	secrets, err := kms.getSecrets()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get secrets for %T: %w", kms,
+			err)
+	}
+
+	err = setConfigString(&kms.accessKey, secrets, awsAccessKey)
+	if err != nil {
+		return nil, err
+	}
+	err = setConfigString(&kms.secretAccessKey, secrets,
+		awsSecretAccessKey)
+	if err != nil {
+		return nil, err
+	}
+	// awsSessionToken is optional
+	err = setConfigString(&kms.sessionToken, secrets, awsSessionToken)
+	if errors.Is(err, errConfigOptionInvalid) {
+		return nil, err
+	}
+
+	return kms, nil
+}
+
+func (kms *AWSMetadataKMS) getSecrets() (map[string]interface{}, error) {
+	c := NewK8sClient()
+	secret, err := c.CoreV1().Secrets(kms.namespace).Get(context.TODO(),
+		kms.secretName, metav1.GetOptions{})
+	if err != nil {
+		return nil, fmt.Errorf("failed to get Secret %s/%s: %w",
+			kms.namespace, kms.secretName, err)
+	}
+
+	config := make(map[string]interface{})
+
+	for k, v := range secret.Data {
+		switch k {
+		case awsSecretAccessKey, awsAccessKey:
+			config[k] = string(v)
+		default:
+			return nil, fmt.Errorf("unsupported option for KMS "+
+				"provider %q: %s", kmsTypeAWSMetadata, k)
+		}
+	}
+
+	return config, nil
+}
+
+func (kms *AWSMetadataKMS) Destroy() {
+	// Nothing to do.
+}
+
+// requiresDEKStore indicates that the DEKs should get stored in the metadata
+// of the volumes. This Amazon KMS provider does not support storing DEKs in
+// AWS as that adds additional costs.
+func (kms *AWSMetadataKMS) requiresDEKStore() DEKStoreType {
+	return DEKStoreMetadata
+}
+
+func (kms *AWSMetadataKMS) getService() (*awsKMS.KMS, error) {
+	creds := awsCreds.NewStaticCredentials(kms.accessKey,
+		kms.secretAccessKey, kms.sessionToken)
+
+	sess, err := awsSession.NewSessionWithOptions(awsSession.Options{
+		SharedConfigState: awsSession.SharedConfigDisable,
+		Config: aws.Config{
+			Credentials: creds,
+			Region:      aws.String(kms.region),
+		},
+	})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create AWS session: %w", err)
+	}
+
+	return awsKMS.New(sess), nil
+}
+
+// EncryptDEK uses the Amazon KMS and the configured CMK to encrypt the DEK.
+func (kms *AWSMetadataKMS) EncryptDEK(volumeID, plainDEK string) (string, error) {
+	svc, err := kms.getService()
+	if err != nil {
+		return "", fmt.Errorf("could not get KMS service: %w", err)
+	}
+
+	result, err := svc.Encrypt(&awsKMS.EncryptInput{
+		KeyId:     aws.String(kms.accessKey),
```

Comment:
This needs to be an ARN for the CMK, which is different from the aws_access_key_id.
There needs to be an additional parameter for the CMK ARN in the secret.

Reply:
@nbalacha so do we need both `cmk` key and `access key` for the encryption or just `cmk` key is enough ?

Reply:
We need both. The cmk is the actual key that will be used to perform the encryption/decryption. The access_key is used as part of the credentials to authenticate the user.

Reply:
@humblec , example here: https://docs.aws.amazon.com/sdk-for-go/v1/developer-guide/kms-example-encrypt-data.html

Reply:
Thanks @nbalacha , it helps :+1:

Reply:
@nixpanic, How do I set up encryption for PVs? From the OCP console it looks like this needs to be set up in the storage class. 
1. Can there be a separate CMK per storageclass ?
2. Will the CMK be an entry in the secret or in something related to the storageclass?

Reply:
Thank @nbalacha! Is there a standard environment variable that is used for the CMK-ARN? Or, can you recommend a variable that you would like to see used?

Reply:
AWS_CMK_ARN or AWS_KEY_ID are possible options.

Reply:
will add a `AWS_CMK_ARN` option